modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/rel/ModifyNode.java

```diff
@@ -197,69 +174,203 @@ private void tryEnd() throws Exception {
+     * clauses, thus has na UPSERT semantic.
+     *
+     * <p>The rows passed to the node have one of the possible format: <ol>
+     *     <li>[insert row type] -- the node is INSERT or MERGE with WHEN NOT MATCHED clause only</li>
+     *     <li>[delete row type] -- the node is DELETE</li>
+     *     <li>[full row type] + [columns to update] -- the node is UPDATE or MERGE with WHEN MATCHED clause only</li>
+     *     <li>[insert row type] + [full row type] + [columns to update] -- the node is MERGE with both handlers, has UPSERT semantic</li>
+     * </ol>
+     *
+     * @param rowSize The size of the row passed to the node.
+     * @return {@code true} if the node is MERGE with UPSERT semantic.
+     */
+    private boolean hasUpsertSemantic(int rowSize) {
+        return mapping != null && updateColumns != null && rowSize > mapping.length + updateColumns.size();
+    }
+
+    /**
+     * Creates a mapping to inline updates into the row.
+     *
+     * <p>The row passed to the modify node contains columns specified by
+     * {@link TableDescriptor#selectForUpdateRowType(IgniteTypeFactory)} followed by {@link #updateColumns}. Here is an example:
+     *
+     * <pre>
+     *     CREATE TABLE t (a INT, b INT, c INT);
+     *     INSERT INTO t VALUES (2, 2, 2);
+     *
+     *     UPDATE t SET b = b + 10, c = c * 10;
+     *     -- If selectForUpdateRowType specifies all the table columns,
+     *     -- then the following row should be passed to ModifyNode:
+     *     -- [2, 2, 2, 12, 20], where first three values is the original values
+     *     -- of columns A, B, and C respectively, 12 is the computed value for column B,
+     *     -- and 20 is the computed value for column C
+     * </pre>
+     *
+     * <p>For example above we should get mapping [0, 3, 4]: <ul>
+     *     <li>The column at index 0 is A. It's not updated, thus it mapped onto itself: 0 --> 0</li>
+     *     <li>The column at index 1 is B. B should be updated, and new value for B is at the index 3: 1 --> 3</li>
+     *     <li>The column at index 2 is C. C should be updated, and new value for C is at the index 4: 2 --> 4</li>
+     * </ul>
+     *
+     * @param descriptor A descriptor of the target table.
+     * @param updateColumns Enumeration of columns to update.
+     * @return A mapping to inline the updates into the row.
+     */
+    private static int @Nullable [] mapping(TableDescriptor descriptor, @Nullable List<String> updateColumns) {
+        if (updateColumns == null) {
+            return null;
+        }
+
+        int columnCount = descriptor.columnsCount();
+
+        int[] mapping = new int[columnCount];
+
+        Object2IntMap<String> updateColumnIndexes = new Object2IntOpenHashMap<>(updateColumns.size());
+
+        for (int i = 0; i < updateColumns.size(); i++) {
+            updateColumnIndexes.put(updateColumns.get(i), i + columnCount);
+        }
```

Comment:
for (int i = 0; i < columnCount; i++) {
    mapping[i] = i;
}

 for (int i = 0; i < updateColumns.size(); i++) {
     String colName = updateColumns.get(i); 
     ColumnDescriptor columnDescriptor = descriptor.columnDescriptor(colName);
     
      mapping[columnDescriptor.logicalIndex()] = columnCount + i;
 }

Reply:
thanks, fixed